src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -746,18 +753,18 @@ public void close(String reason, boolean flushEngine) throws IOException {
                 }
                 changeState(IndexShardState.CLOSED, reason);
             } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
```

Comment:
I don't think we should do this. We are not ref-coutning the engine then we would do it on the engine. The engine should be close immediately on the close call in the IndexShard. This PR should only add the refCounting and the ability to wait until all references are returned. the purpose of this is to prevent crypting errors from the doc mapper etc. when indexing since we retry on EngineClosed exceptions the user won't see them.